internal/antewrapper/prov_feegrant_test.go

```diff
@@ -0,0 +1,234 @@
+package antewrapper_test
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/tendermint/tendermint/crypto"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/codec"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+
+	"github.com/cosmos/cosmos-sdk/simapp"
+	"github.com/cosmos/cosmos-sdk/simapp/helpers"
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/simulation"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authsign "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	"github.com/cosmos/cosmos-sdk/x/auth/tx"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/cosmos/cosmos-sdk/x/feegrant"
+	pioante "github.com/provenance-io/provenance/internal/antewrapper"
+)
+
+func (suite *AnteTestSuite) TestDeductFeesNoDelegation() {
+	suite.SetupTest(false)
+	// setup
+	app, ctx := suite.app, suite.ctx
+
+	protoTxCfg := tx.NewTxConfig(codec.NewProtoCodec(app.InterfaceRegistry()), tx.DefaultSignModes)
+
+	dfd := pioante.NewProvenanceDeductFeeDecorator(app.AccountKeeper, app.BankKeeper, app.FeeGrantKeeper, app.MsgFeesKeeper)
+
+	// this just tests our handler
+	decorators := []sdk.AnteDecorator{pioante.NewFeeMeterContextDecorator(), dfd}
+
+	feeAnteHandler := sdk.ChainAnteDecorators(decorators...)
+
+	// this tests the whole stack
+	anteHandlerStack := suite.anteHandler
+
+	// keys and addresses
+	priv1, _, addr1 := testdata.KeyTestPubAddr()
+	priv2, _, addr2 := testdata.KeyTestPubAddr()
+	priv3, _, addr3 := testdata.KeyTestPubAddr()
+	priv4, _, addr4 := testdata.KeyTestPubAddr()
+	priv5, _, addr5 := testdata.KeyTestPubAddr()
+
+	// Set addr1 with insufficient funds
+	err := simapp.FundAccount(suite.app.BankKeeper, suite.ctx, addr1, []sdk.Coin{sdk.NewCoin("atom", sdk.NewInt(10))})
+	suite.Require().NoError(err)
+
+	// Set addr2 with more funds
+	err = simapp.FundAccount(suite.app.BankKeeper, suite.ctx, addr2, []sdk.Coin{sdk.NewCoin("atom", sdk.NewInt(99999))})
+	suite.Require().NoError(err)
+
+	// grant fee allowance from `addr2` to `addr3` (plenty to pay)
+	err = app.FeeGrantKeeper.GrantAllowance(ctx, addr2, addr3, &feegrant.BasicAllowance{
+		SpendLimit: sdk.NewCoins(sdk.NewInt64Coin("atom", 500)),
+	})
+	suite.Require().NoError(err)
+
+	// grant low fee allowance (20atom), to check the tx requesting more than allowed.
+	err = app.FeeGrantKeeper.GrantAllowance(ctx, addr2, addr4, &feegrant.BasicAllowance{
+		SpendLimit: sdk.NewCoins(sdk.NewInt64Coin("atom", 20)),
+	})
+	suite.Require().NoError(err)
+
+	cases := map[string]struct {
```

Comment:
i agree on the non determinism part of maps, but just test cases mapped to test, so i think it's ok here imo.

Reply:
It makes them harder to find when they fail. It also can cause issues where tests sometimes fail because sometimes one test goes first and changes something that another test has a problem with. We've run into that before in this repo.

Reply:
ok changed to array..
i would say it should be a bad thing if they failed if they were run in a different order, cause tests should pass irrespective of order, but i changed it 🤷  agreed they are harder to debug but also them breaking because of order is code smell or test code smell at the very least.